### PR TITLE
DBZ-409 SMT for converting MongoDB CDC events to flat structure

### PR DESCRIPTION
**Pull request by sairam881990**
https://issues.jboss.org/browse/DBZ-409

### CODE REVIEW

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import org.junit.Test;
+
+import io.debezium.data.Envelope.FieldName;
+import io.debezium.data.Json;
+/**
+ * Unit test for {@code StringToJSON}.
+ *
+ * @author Sairam Polavarapu
+ */
+public class UnwrapFromMongoDbEnvelopeTest {
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java]: Tbh. I don't think this test makes too much sense. Most notably, it doesn't invoke the SMT `UnwrapFromMongoDbEnvelope` at all. It seems more like it re-implements it? It's not quite clear how this supposed to work. I'll push an example for how it should look like in a bit.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+/**
+ * Debezium Mongo Connector generates the CDC records in String format. Sink connectors usually are not able to parse
+ * the string and insert the document as it is represented in the Source. so a user use this SMT to parse the String
+ * and insert the MongoDB document in the JSON format.
+ *
+ * @param <R> the subtype of {@link ConnectRecord} on which this transformation will operate
+ * @author Sairam Polavarapu
+ */
+public class UnwrapFromMongoDbEnvelope<R extends ConnectRecord<R>> implements Transformation<R> {
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: Sigh, this still emits the "_id" field when using PK mode "record_key" :( As discussed on the other PR, this shouldn't be the case. I'll get rid of it.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/MongoDataConverter.java [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+            SchemaBuilder builderDoc = SchemaBuilder.struct();
+            BsonDocument docs = keyValuesforSchema.getValue().asDocument();
+
+            for (Entry<String, BsonValue> doc : docs.entrySet()) {
+                addFieldSchema(doc, builderDoc);
+            }
+            builder.field(key, builderDoc.build());
+            break;
+
+        case ARRAY:
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/MongoDataConverter.java]: Will this work with a deeply nested structure? I.e. an array of arrays? I think this ought to contain some kind of recursiveness really.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: Ok, that's fixed with the commits I added.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/MongoDataConverter.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/MongoDataConverter.java]: I've filed https://issues.jboss.org/browse/DBZ-562 as a follow-up for that one.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java]: Test is largely re-written now as per my update.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: To be more precise, "_id" is never part of the emitted structure now, instead always and only the "id" field(s) from the message key will be emitted.

**Comment by gunnarmorling** [comment]
@sairam881990 so I've pushed a couple of more commits to this PR, with that I think we can release it as "preview-quality" feature. The last remaining issue is to describe the new SMT on the website. Would you fancy sending a PR for creating a new page similar to [this one](https://github.com/debezium/debezium.github.io/blob/develop/docs/configuration/event-flattening.asciidoc)? Thanks!

Also I've filed three follow-up tasks in JIRA:

* [DBZ-561](https://issues.jboss.org/browse/DBZ-561) Provide option to flatten structs in MongoDB unwrapping SMT 
* [DBZ-562](https://issues.jboss.org/browse/DBZ-562) MongoDB retrieval SMT: add support for arrays of arrays 
* [DBZ-563](https://issues.jboss.org/browse/DBZ-563) Add option for dropping deletes and tombstone events to MongoDB struct recreation SMT 

I think we can deliver them in subsequent releases. If you'd like to work on any of them, please let me know - any help is welcomed! Esp. DBZ-561 and DBZ-563 are important to make this work with existing "flat format" sink connectors, most notably the JDBC sink connector.

@jpechane Could you take a final look, too, and merge it if you don't see any further issues? Thanks!

**Comment by sairam881990** [comment]
Sure. I will create the SMT document and sending a PR.

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks a lot, @sairam881990! In case you're up for any of the follow-ups (esp. DBZ-561 and DBZ-563) please let us know.

**Pull request merged**